### PR TITLE
wrapped data loader to make it error prove

### PR DESCRIPTION
**Pull request by milamer**
## Description

I changed the dataloader to be wrapped in our own wrapper to ensure:
- no recursion
- no wrong keys
- Possible to use object instead of string

Used the undocumented feature of the dataloader to return an array of Promises instead of one Promise

Added timeout to dataloader batch to reject all promises after 30s

Added recursion check

Changed richtext population from payload.findById to dataloader.load

- [x] I have read and understand the CONTRIBUTING.md document in this repository

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Existing test suite passes locally with my changes

### CODE REVIEW

File: src/collections/dataloader.ts [hunk 1]
```diff
@@ -51,63 +47,160 @@ const batchAndLoadDocs = (req: PayloadRequest): BatchLoadFn<string, TypeWithID>
...
   // Run find requests in parallel
+  Object.entries(batchByFindArgs).forEach(async ([batchKey, ids]) => {
+    const allKeysInBatch = new Set<string>();
+    ids.forEach(({ key }) => allKeysInBatch.add(key));
+
+    try {
+      const [collection, depth, currentDepth, locale, fallbackLocale, overrideAccess, showHiddenFields] = JSON.parse(batchKey);
+
+      // copy req object to avoid mutating it
+      const request = { ...req };
```
Comment (milamer) [inline src/collections/dataloader.ts:78]: Are there other ways to copy a request?

File: src/graphql/schema/buildObjectType.ts [hunk 2]
```diff
@@ -346,16 +349,18 @@ function buildObjectType({
...
-                0,
-                0,
-                locale,
-                fallbackLocale,
-                false,
-                false,
-              ]));
+              const result = await (context.req as PayloadRequest).payloadDataLoader.load(
+                {
+                  collection: collectionSlug as string,
```
Comment (milamer) [inline src/graphql/schema/buildObjectType.ts:354]: type is string | string[]

Can this be fixed?

File: src/graphql/schema/buildObjectType.ts [hunk 3]
```diff
@@ -391,16 +396,18 @@ function buildObjectType({
...
-              0,
-              0,
-              locale,
-              fallbackLocale,
-              false,
-              false,
-            ]));
+            const relatedDocument = await (context.req as PayloadRequest).payloadDataLoader.load(
+              {
+                collection: relatedCollectionSlug as string,
```
Comment (milamer) [inline src/graphql/schema/buildObjectType.ts:401]: type is string | string[]

Can this be fixed?

File: src/fields/hooks/afterRead/relationshipPopulationPromise.ts [hunk 4]
```diff
@@ -40,16 +40,18 @@ const populate = async ({
...
-        currentDepth + 1,
-        req.locale,
-        req.fallbackLocale,
-        overrideAccess,
-        showHiddenFields,
-      ]));
+      relationshipValue = await req.payloadDataLoader.load(
+        {
+          collection: relatedCollection.config.slug,
+          id: id as string,
```
Comment (milamer) [inline src/fields/hooks/afterRead/relationshipPopulationPromise.ts:46]: type is unknown

Can this be fixed?

File: src/collections/dataloader.ts (hunk 1 above)
Comment (jmikrut) [reply src/collections/dataloader.ts:78]: Unfortunately Node 15 introduces some issues with copying a `req`, and we had to work around that in our local operations. I think we might need to work around that here as well, otherwise the `req` that is passed to the `find` operation will lose properties in newer Node versions.

Take a look here:
https://github.com/payloadcms/payload/pull/975

Without being familiar with the exact reasons why we need to avoid mutating `req`, it's hard for me to understand the ramifications of not taking a copy of it.

What do you think?

File: src/graphql/schema/buildObjectType.ts (hunk 3 above)
Comment (jmikrut) [reply src/graphql/schema/buildObjectType.ts:401]: Fixed in https://github.com/payloadcms/payload/tree/fix/dataloader-deadlock

File: src/graphql/schema/buildObjectType.ts (hunk 2 above)
Comment (jmikrut) [reply src/graphql/schema/buildObjectType.ts:354]: Yep - fixed here https://github.com/payloadcms/payload/tree/fix/dataloader-deadlock. Might have to pull this into your branch as I pushed, but my commit didn't show up in your original PR!

File: src/fields/hooks/afterRead/relationshipPopulationPromise.ts (hunk 4 above)
Comment (jmikrut) [reply src/fields/hooks/afterRead/relationshipPopulationPromise.ts:46]: This one is gonna be trickier to fix. I think we should avoid this one for now.

**Comment by jmikrut** [comment]
Hey @milamer I apologize for the delay in getting this PR merged. I believe we have successfully rectified the original issue here (recursion), and we are actually implementing transactions within the dataloader as well as moving to a monorepo structure where our file hierarchy within the Payload repo is going to change. 

In addition, even prior to this PR, this function is quite complex - and the team and I have decided to keep the code as simple as possible in an effort to make it as understandable as possible to the widest audience.

Basically, there are lots of changes in-flight that have overlap with this PR. I am going to close this, but I really really appreciate your help here and your ability to dive so deeply into such a crucial part of Payload.

I hope you are well!
